### ISSUE TITLE
BigTable: mutate_rows is slow.

### ISSUE THREAD

**Issue by mayurjain0312** [GoogleCloudPlatform/google-cloud-python#4713]
Ubuntu 16.04
Python version and virtual environment information python --version - python 2.7.12

with a batch size of 300, total 3 nodes in an instance, the write throughput is not good for bigTable by using mutate_rows api call.

        BULK_WRITE_BATCH_SIZE = 300
        with open(file_path) as sensor_data_input_file:
            list_direct_row_obj = []
            for line in sensor_data_input_file:
                if not line.strip():
                    continue

                sensor_json_data = json.loads(line)
                row_key = create_sensor_data_id(sensor_json_data)
                value = line

                direct_row_obj = bigtable.row.DirectRow(row_key, table)

                column_id = 'column_id_data'.encode('utf-8')
                direct_row_obj.set_cell(column_family_id, column_id, value.encode('utf-8'))

                list_direct_row_obj.append(direct_row_obj)
                direct_row_obj = None

                if len(list_direct_row_obj) == BULK_WRITE_BATCH_SIZE:
                    table.mutate_rows(list_direct_row_obj)
                    list_direct_row_obj[:] = []

            if list_direct_row_obj:
                table.mutate_rows(list_direct_row_obj)

**Comment by sduskis** [comment]
Can you please describe some of the symptoms you're experiencing?  `table.mutate_rows` is a synchronous operation, so you're limited to one call at a time.  How many rows per second are you seeing on the cloud console?

**Comment by mayurjain0312** [comment]
Is this the only "bulk write" option that BigTable offers?
From the console, I can see around 800 rows written per second.
(Production instance with 3 nodes)

**Comment by mayurjain0312** [comment]
We are currently using mongodb as our no-sql database and we would prefer moving to BigTable. As far as the speed of writing data to database is concerned, I see that mongodb is far exceeding BigTable(unless I am not using BigTable in the right manner)

**Comment by mayurjain0312** [comment]
@sduskis: any updates?

**Comment by sduskis** [comment]
A 3 node Cloud Bigtable cluster can handle writing at least 30K rows per second, and each additional node adds 10K rows per second. [ We tested up to 3,500 nodes with Cloud Bigtable](https://cloudplatform.googleblog.com/2016/03/financial-services-firm-processes-25-billion-stock-market-events-per-hour-with-Google-Cloud-Bigtable.html) with a linear increase in performance.

The 800 rows per second is a limitation of the python client and what ever VM you're on.  The process sends 300 rows, and waits for a response.  You can run multiple processes that each read in a different file to improve total throughput against your cluster.

The Java client has an async client that performs far better in these situations, since it has a robust async client.  We have not ported that functionality to python yet.  [Here is an example](https://github.com/GoogleCloudPlatform/cloud-bigtable-examples/tree/master/java/simple-performance-test) with the Java version.

**Comment by mayurjain0312** [comment]
@sduskis : Thanks for responding.
Do you know when can this async functionality for Python be ready?

**Comment by sduskis** [comment]
We don't have an ETA for this functionality in Python.

**Comment by basphilippus** [comment]
Hi,

Is there any progress on this feature? I would really like to scale my writes to Bigtable with Python code.

**Comment by sduskis** [comment]
We still do not have concrete plans to implement this feature.

**Comment by mholiv** [comment]
Hello! Are there any updates on plans to implement this feature?

Async is quickly becoming the norm in terms of modern python development. :)

BigTable IO is consistently the only blocking IO I encounter on a regular basis.

DynamoDB as a async wrapper in the form of aioboto3

**Comment by cwbeitel** [comment]
@sduskis I can help implement this if you can provide me a little guidance. Can provide a rough sketch?

**Comment by sduskis** [comment]
@cwbeitel  There's a lot that goes into this.  There's a Java implementation that would be similar, but I'm not sure how much it would help ([here](https://github.com/googleapis/google-cloud-java/blob/master/google-cloud-clients/google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/models/BulkMutationBatcher.java)).

Here are some constraints that I think are important:

1. Allow up to 4 concurrent RPCs in progress, after that block.
2. Each RPC batch should consist of mutations that meet one of the max criteria:
   1. 125 rows
   2. 10,000 total "entries" / column qualifier updates across all rows
   3. an RPC size of no more than 5 MB
3. The interface should have the following methods:
   1. **b = table.createBatcher()** - creates a new multi-RPC batcher
   2. **b.add(mutation)** - Blocks, if necessary.  Adds the row an in - memory buffer.  Send an RPC if the buffer is full.
   3. **b.flush()** - sends out the in-memory buffer and waits for all RPCs to complete.

**Comment by cwbeitel** [comment]
@sduskis Yeah that's cool I hear those thoughts. Thanks for sharing the Java implementation, that's helpful. Looks like the PubSub python client [code](https://github.com/googleapis/google-cloud-python/blob/master/pubsub/google/cloud/pubsub_v1), e.g. the [published message batcher](https://github.com/googleapis/google-cloud-python/blob/master/pubsub/google/cloud/pubsub_v1/publisher/_batch/base.py), would also be helpful to emulate.

Here's a [gist](https://gist.github.com/cwbeitel/242ab97241d8d6e59c81139a1e4f4142) contextualizing this in the case of streaming deep learning training examples to a cbt table.

Having had a look at this I'm probably going to first try to accomplish the same using the Golang client given how involved it would be to do this with the Python client.

**Comment by peschue** [comment]
Any updates since last year?
